src/boot/efi/boot.c

```diff
@@ -544,21 +544,16 @@ static BOOLEAN menu_run(
         else
                 timeout_remain = -1;
 
-        idx_highlight = config->idx_default;
-        idx_highlight_prev = 0;
-
         visible_max = y_max - 2;
 
-        if ((UINTN)config->idx_default >= visible_max)
-                idx_first = config->idx_default-1;
-        else
+        if (config->entry_count < visible_max || idx_highlight < visible_max / 2)
                 idx_first = 0;
-
+        else if (idx_highlight >= config->entry_count - (visible_max / 2))
+                idx_first = config->entry_count - visible_max;
+        else
+                idx_first = idx_highlight - (visible_max / 2);
```

Comment:
hmm, so what are you precisely changing here? can you explain? (maybe add comments to the sources, explaining that)